tla-types/src/main/scala/at/forsyte/apalache/tla/typedbuilder/TT1OperatorSignatureMatcher.scala

```diff
@@ -0,0 +1,819 @@
+  }
+
+  // \A T . (Set(Set(T))) => Set(T)
+  def matchBigUnionSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(SetT1(SetT1(t))),
+          SetT1(t)
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Set(T)) => Bool
+  def matchSetPredicateSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(SetT1(t)),
+          BoolT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Set(T)) => Int
+  def matchCardinalitySig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(SetT1(t)),
+          IntT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (T) => T
+  def matchEndomorphismSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(t),
+          t
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Bool,T) => Bool
+  def matchStutterSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(BoolT1(), t),
+          BoolT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (T,Bool) => Bool
+  def matchTemporalWithVarSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(t, BoolT1()),
+          BoolT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T_1,T_2 . (Set(T_1), Set(T_2)) => Set(T_1 -> T2)
+  def matchFunSetSig(args: Seq[TlaEx]): SignatureMatch = {
+    val Seq(t1, t2) = typeVarGenerator.getNUnique(2)
+    val operT =
+      OperT1(
+          Seq(SetT1(t1), SetT1(t2)),
+          SetT1(FunT1(t1, t2))
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T,T_1,...,T_n . (T, T_1, Set(T_1),..., T_m, Set(T_m)) =>
+  // (case n = 1) : T_1 -> T
+  // (case n > 1) : (T_1, ..., T_n) -> T
+  def matchFunDefSig(args: Seq[TlaEx]): SignatureMatch = {
+    val nArgs = args.length
+    assert(nArgs >= 3 && nArgs % 2 == 1)
+    val n = (nArgs - 1) / 2
+    val t = typeVarGenerator.getUnique
+    val ts = typeVarGenerator.getNUnique(n)
+    val allPairs = ts flatMap { x => Seq(x, SetT1(x)) }
+    val domT = ts match {
+      case single +: Nil => single
+      case _             => TupT1(ts: _*)
+    }
+    val operT =
+      OperT1(
+          t +: allPairs,
+          FunT1(domT, t)
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T_1,...,T_n . (Str, T_1,..., Str, T_m) => [ s1: T_1, ... , sn: T_n ]
+  def matchRecEnumSig(args: Seq[TlaEx]): SignatureMatch = {
+    val recordKeyTypePairs = args.zipWithIndex.collect {
+      case (ValEx(TlaStr(s)), i) if i % 2 == 0 =>
+        s -> typeVarGenerator.getUnique
+    }
+
+    val argTypes = recordKeyTypePairs flatMap { case (_, typeVar) =>
+      Seq(StrT1(), typeVar)
+    }
```

Comment:
Arguably, this could be more declarative with [`grouped`](https://www.scala-lang.org/api/current/scala/collection/Seq.html#grouped(size:Int):Iterator[C]). As in https://github.com/informalsystems/apalache/blob/unstable/tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala#L314-L319 -- It's maybe a matter of taste tho, so just a weak suggestion here.